src/ray/core_worker/core_worker.cc

```diff
@@ -1760,36 +1760,30 @@ void CoreWorker::HandleDirectActorCallArgWaitComplete(
 void CoreWorker::HandleGetObjectStatus(const rpc::GetObjectStatusRequest &request,
                                        rpc::GetObjectStatusReply *reply,
                                        rpc::SendReplyCallback send_reply_callback) {
+  if (HandleWrongRecipient(WorkerID::FromBinary(request.owner_id()),
+                           send_reply_callback)) {
+    return;
+  }
+
   ObjectID object_id = ObjectID::FromBinary(request.object_id());
   RAY_LOG(DEBUG) << "Received GetObjectStatus " << object_id;
-  TaskID owner_id = TaskID::FromBinary(request.owner_id());
-  if (owner_id != GetCallerId()) {
-    RAY_LOG(INFO) << "Handling GetObjectStatus for object produced by previous task "
-                  << owner_id.Hex();
-  }
-  // We own the task. Reply back to the borrower once the object has been
-  // created.
-  // TODO(swang): We could probably just send the object value if it is small
-  // enough and we have it local.
-  reply->set_status(rpc::GetObjectStatusReply::CREATED);
-  if (task_manager_->IsTaskPending(object_id.TaskId())) {
-    // Acquire a reference and retry. This prevents the object from being
-    // evicted out from under us before we can start the get.
-    AddLocalReference(object_id, "<temporary (get object status)>");
-    if (task_manager_->IsTaskPending(object_id.TaskId())) {
-      // The task is pending. Send the reply once the task finishes.
-      memory_store_->GetAsync(object_id,
-                              [send_reply_callback](std::shared_ptr<RayObject> obj) {
-                                send_reply_callback(Status::OK(), nullptr, nullptr);
-                              });
-      RemoveLocalReference(object_id);
-    } else {
-      // We lost the race, the task is done.
-      RemoveLocalReference(object_id);
-      send_reply_callback(Status::OK(), nullptr, nullptr);
-    }
+  bool in_scope = reference_counter_->AddBorrower(object_id, request.borrower_address());
+  if (in_scope) {
```

Comment:
I like this change. Just want to verify this isn't an issue though: if between this line and the previous, the borrower was removed by a concurrent RPC, would we hang trying to fetch from the memory store?

Reply:
Oh nice catch, I think that would be an issue. I guess for now we can do something similar to before where we add a local ref during `AddBorrower`, but it's pretty hacky.

I think it could make sense eventually to merge the memory_store_ with the reference_counter_. All these possible race conditions are a bit concerning.